[Checkout 2.1.0] Add AddressValidation container documentation and tutorial (#443)
commerce-docs/microsite-commerce-storefront: src/content/docs/dropins/checkout/containers/address-validation.mdx

---
title: AddressValidation container
description: Configure the AddressValidation container to present suggested vs. original shipping addresses and act on the shopper's choice.
---

import OptionsTable from '@components/OptionsTable.astro';
import Aside from '@components/Aside.astro';

> The `AddressValidation` container displays a suggested shipping address (for example, returned by a third‑party address verification service) alongside the currently entered address, and lets the shopper choose which one to use.

Reviewer: ```suggestion
The `AddressValidation` container displays a suggested shipping address (from a third-party verification service) alongside the entered address, allowing shoppers to choose between them.
```

> Typical usage is to invoke this container from a modal during the place‑order flow after you call your own address verification service.

Reviewer: ```suggestion
Typically invoked from a modal during checkout after calling your address verification service.
```

## AddressValidation configurations

The `AddressValidation` container provides the following configuration options:

<OptionsTable
  compact
  options={[
    ['Option', 'Type', 'Req?', 'Description'],
    ['suggestedAddress', 'CartAddressInput | null', 'No', 'Address suggestion to present to the shopper.'],
    ['handleSelectedAddress', 'function', 'No', 'Async callback fired when the shopper selects an address. Receives the selection and the chosen address.'],
  ]}
/>

### AddressValidationProps interface

The `AddressValidation` container receives an object that implements the following interface:

```ts
interface AddressValidationProps {
  suggestedAddress: Partial<CartAddressInput> | null;
  handleSelectedAddress?: (payload: {
    selection: 'suggested' | 'original';
    address: CartAddressInput | null | undefined;
  }) => void;
}
```

> - `suggestedAddress` is the normalized address to propose to the shopper.

Reviewer: ```suggestion
- `suggestedAddress` - The normalized address to propose to the shopper.
```

> - `handleSelectedAddress` is called when the shopper chooses either the suggested or original address. Use it to persist the selection or continue the checkout flow.

Reviewer: ```suggestion
- `handleSelectedAddress` - Called when the shopper selects an address. Use to persist the selection or continue checkout.
```

## CartAddressInput type

> The `CartAddressInput` type used by this container has the following shape:

Reviewer: ```suggestion
The `CartAddressInput` type has this shape:
```

```ts
interface CartAddressInput {
  city: string;
  countryCode: string;
  postcode: string;
  region: string;
  street: string[];
}
```

<Aside type="tip" title="Get the current address">

> The container automatically listens to checkout events and maps the current shipping address into a `CartAddressInput`. You only need to pass the `suggestedAddress` when you have one.

Reviewer: ```suggestion
The container listens to checkout events and automatically maps the current shipping address to `CartAddressInput`. Only pass `suggestedAddress` when available.
```

</Aside>

<Aside type="note" title="Normalization">

> Ensure your address verification service output is transformed to `CartAddressInput` (fields such as `street`, `city`, `region`, `countryCode` or `postcode`) before passing it to the container. If a property is not provided, the default data from the original address will be used.

Reviewer: ```suggestion
Transform your address verification service output to `CartAddressInput` format (with fields like `street`, `city`, `region`, `countryCode`, `postcode`) before passing to the container. Missing properties default to the original address values.
```

</Aside>

## Example

> For a complete integration walkthrough, see the tutorial [Validate shipping address](/dropins/checkout/tutorials/validate-shipping-address/).

Reviewer: ```suggestion
 For a complete walkthrough, see the [Validate shipping address](/dropins/checkout/tutorials/validate-shipping-address/) tutorial.
```
